src/server/socket.js

```diff
@@ -28,20 +44,112 @@ function SocketServer(simulatorProxy, project) {
     this._pendingEmits[SIM_HOST] = [];
     this._pendingEmits[DEBUG_HOST] = [];
 
-    /* Deferred are used to delay the actions until an event happens.
-     * Two major events are kept track of:
-     *      - When app-host connects (sends initial message)
-     *      - When sim-host notifies it's ready to serve an app-host.
-     */
-    this._whenAppHostConnected = Q.defer();
-    this._whenSimHostReady     = Q.defer();
+    // current client states
+    this._ahState = AH_STATE_DISCONNECTED;
+    this._shState = SH_STATE_DISCONNECTED;
 
     var config = this._simulatorProxy.config,
         telemetry = this._simulatorProxy.telemetry;
 
     this._liveReload = new LiveReload(project, telemetry, config.forcePrepare);
 }
 
+/**
+ * Handles protocol events triggered by the change of states in one of the
+ * clients (APP_HOST or SIM_HOST). Protocol events are those which depend on
+ * both APP_HOST and SIM_HOST.
+ */
+SocketServer.prototype._handleStateChange = function (appHostState, simHostState) {
+    var state = appHostState | simHostState;
+    switch (state) {
+        case AH_STATE_CONNECTED | SH_STATE_READY:
+            this._setupAppHostHandlers();
+            this._setupSimHostHandlers();
+            this._subscribeTo(APP_HOST, 'app-plugin-list', function (data) {
+                this._emitTo(SIM_HOST, 'app-plugin-list', data);
+                this._setAppHostState(AH_STATE_PLUGINS_READY);
+            }.bind(this), true);
+            this._emitTo(APP_HOST, 'init');
+            this._setAppHostState(AH_STATE_INIT_SENT);
+            break;
+        case AH_STATE_PLUGINS_READY | SH_STATE_READY:
+            this._subscribeTo(SIM_HOST, 'start', function () {
```

Comment:
No, it's not possible by the [protocol design](https://github.com/Microsoft/cordova-simulate/wiki/The-cordova-simulate-bootstrap-protocol) and by the implementation (to your specific point, `init` will never be sent to `APP_HOST` and `APP_HOST` will not send anything proactively until it's received `init`).

I was not probably explicit, but this code implements a spec. It is a built-in design point, though, that either of the clients could connect at any point in time. Could you review the [spec](https://github.com/Microsoft/cordova-simulate/wiki/The-cordova-simulate-bootstrap-protocol) and tell me if it's not detailed enough or if it has a flaw? It could use a fresh look. If the protocol is ok, we could then discuss the implementation. If it's flawed, it has to be fixed.